### DISCUSSION TITLE
Server starts, but unable to connect Browser

### DISCUSSION THREAD

**Discussion by valhuber** [community/community#17984] [Codespaces] [answered]
Exploring codespaces for first time... my [project](https://github.com/valhuber/Tutorial-ApiLogicProject) (a webserver - api and react admin app) loaded and ran beautifully, but unable to open browser on running app.

When I run this app _locally_ with VSCode, I accept the recommendation for docker, and the server runs (just as it does in codespaces).  A pre-created Launch config starts the server, also just as it does in codespaces.

It differs in that locally, VSCode detects it as a web server, and offers to start the simple Browser.  This does not occur in code spaces.

I tried to start the simple browser, and supplying the url recommended in the readme (http://localhost:5656/admin-app/index.html#/Home), but it fails to load.

I should note that the project intentionally uses port 5656 (not localhost), to avoid port conflicts with other local projects.

**Comment by joshspicer** [comment, marked as answer]
Hey @valhuber!  I think if I understand your question correctly, you're saying that you were able to clone this repo and start the web server via the VS Code launch config, but weren't successful in Codespaces?

I followed the following steps:
1. Created codespace from the project link you supplied
2. Launched the app with the vscode `ApiLogicServer` run config.

I think I was able to accomplish what you were trying.  Note that VS Code picked up on the running port and exposed it here in the `Ports` tab. If it doesn't show up for you, you can click `Add Port` and type `5656`.

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/23246594/172220798-f1928288-7be5-4b14-a52e-3621449efda5.png">

As an aside, I did notice that your custom Docker image was missing a couple tools that I think may potentially be impacting your experience, notably `git` is missing. We offer a prebuilt image `http://mcr.microsoft.com/vscode/devcontainers/python` that we think works great with Codespaces, if you want to avoid needing to build your own image.

**Comment by valhuber** [comment]
Hi, Josh

Thanks so much for getting back so quickly.  Much appreciated.

Yes, the app runs if I add the port.  So thanks for that.

And, I followed your advice to include git in the image.

I cannot use a standard one since this app is built by and uses ApiLogicServer <https://github.com/valhuber/ApiLogicServer/blob/main/README.md> - it creates projects, like this one, by database introspection, here a version of Northwind.

My interest in Codespaces is partly curiosity (an amazing idea), but mostly product-focused.  ApiLogicServer is pretty easy to install, once you have Python.  But that’s the rub.  Among my trials, Python install failures and glitches have been the norm.

So, Codespaces looks like a welcome way out - no install, just start developing.  Very promising.

If you are interested, I would like to engage a bit in discussing codespaces for generators that create projects.  This may not be a critical use case for you, but do consider how nice it would be to have an instant running api & ui for any sql database, running in the cloud, no install.

I am ***@***.*** ***@***.***>.  My schedule is busy today and Wed, but open afterwards.  Weekends are fine if that is easier for you.  I am in the Bay Area.

Cheers, and thanks again,
Val
